Facelift FFmpeg code quality

illuusio opened the pull request:
There have been several function deprecations since FFmpeg version 1.1. Those generate warning so remove them with #ifdef-them out.
Xiph family (Vorbis, Opus, FLAC and probably others) seeking backward wasn't working as expected so fix that on working towards not crashing test suite.
This is for head (1.13) but those OGG seeking things should be also applied in version 1.12.

illuusio committed:
Remove header license files to favor LICENSE file

illuusio committed:
New FFMpeg/AVConv depricates new functions. #ifdef them to support old and new

illuusio committed:
Fixed seek end and then seek near to start error

illuusio committed:
Most of OGG seeking issues are ironed out. Now it's possible to seek from end to start with rev-button without hickups

illuusio committed:
Code styling update

illuusio committed:
Fixed two major seeking issues that plagues in same time.

These made looping nearly imposible task. If you started playing from the begining or you had CUE point before loop.
    1# Issue was OGG have (and MP4 also most MP3) pre-frame which is 0 < PTS which means it's not mentioned to be played.
       Discarding that because somewho it's imposible to read from some OGG files first frame. Also turn FFmpeg byte offset
       to Frames when seeking starting point. Faster and easier to use with new system.
    2# Miss understanding how 'avformat_seek_file' works. Not PTS is at begining not at the some random Frame from the start.
       So loops are exact and starts always from same point.

Only big problem with this is that Loops are now in wrong place if they were made with older FFmpeg implementation.

illuusio committed:
Added more comments why some things are done like they are

illuusio commented:
Hopefully i noticed all comments and other commit makes FFmpeg soundsource equal to others.

illuusio committed:
FFmpeg way of closing and releasing memory has changed. Finally changed also code to follow this. Also check out of memory situation with alloc

illuusio committed:
More checks if memory has runned out. Also tuned everything to use CSAMPLE aka float pointer

illuusio committed:
Code formating for fitting in Mixxx

illuusio committed:
Now FFmpeg obey channel count to make test not crash.

illuusio commented:
Ok now all test should run but they don't pass so that's short term mission after this.

illuusio commented:
Ok it was my bad.. all tests that are related to SoundSource passes for me. So this is read to be pushed to master from me. I'll have port these to 1.12 branch so it'll be also in the best shape.

illuusio committed:
It's bad habit to use #define and it could be done with method but someway this is cleaner to me

daschuer commented on src/sources/soundsourceffmpeg.cpp:
```diff
@@ -451,16 +512,16 @@ SINT SoundSourceFFmpeg::seekSampleFrame(SINT frameIndex) {
                                  0,
-                                 32767 * 2,
+                                 0,
                                  32767 * 2,
```
What is this magic number? Does a hex notation make it more readable 0x7fff?

daschuer commented on src/sources/soundsourceffmpeg.cpp:
```diff
@@ -490,7 +560,7 @@ SINT SoundSourceFFmpeg::seekSampleFrame(SINT frameIndex) {
             readFramesToCache((AUDIOSOURCEFFMPEG_CACHESIZE - 50), -1);
         } else {
-            readFramesToCache((AUDIOSOURCEFFMPEG_CACHESIZE / 2), frameIndex);
+            readFramesToCache(100, frameIndex);
```
Can you explain 100?

daschuer commented on src/sources/soundsourceffmpeg.cpp:
```diff
@@ -364,39 +403,61 @@ bool SoundSourceFFmpeg::readFramesToCache(unsigned int count, SINT offset) {
-            if (l_SObj == NULL || (l_lPos + 5) > (unsigned int)m_SCache.size()) {
+            if (l_SObj == NULL || (l_lPos + 5) > m_SCache.size()) {
                 offset = l_SObj->startFrame;
                 if (readFramesToCache(50, -1) == false) {
```
What does -1 do here? Offset == -1?  Offset from  current position and -1 is special? 
Can a #define or a a wrapper function help?

daschuer commented on src/sources/soundsourceffmpeg.cpp:
```diff
@@ -364,39 +403,61 @@ bool SoundSourceFFmpeg::readFramesToCache(unsigned int count, SINT offset) {
-        l_lLeft = (size * sizeof(CSAMPLE)) * 2;
-        memset(buffer, 0x00, l_lLeft);
+        l_lLeft = size * AUDIOSOURCEFFMPEG_MIXXXFRAME_TO_BYTEOFFSET;
+        memset(l_pBuffer, 0x00, l_lLeft);
```
this seams to be redundant.

illuusio replied:
If you mean memset. Yes it's very redundant. Just an habit to clean everything. Is you mean l_lLeft then you incorrect.

daschuer replied:
Yes memset()

illuusio committed:
Added more explanation why some numbers are chosen and explantios about how things work

illuusio committed:
Stilization of comment and remove memset that is redundant

illuusio commented:
Ok more cleaning of comments and adding some explanations why some are chosen.

illuusio committed:
More accurate framecount for MP3 and all

illuusio committed:
Few error messages to confuse people even more

illuusio commented:
Now heavy testing phase..

illuusio committed:
Because of length calculation is sometimes incorrect we must fill buffer with 0x00 (zero) to prevent ugly clicks at the end

illuusio committed:
More precise was more non precise so preventing and now it's really more precises time

illuusio committed:
If file end and soundsource thinks we have to still cache some handle that like it should. after this application doesn't make silent caps at the end of file if music dies very quickly at the end

illuusio committed:
Tune what file ends we can handle. If we can handle 'm4a' we can handle 'mp4' also. WMA is 'xwma' so fix that recognization too

illuusio committed:
Now WMA playing is working. WMA has huge FFmpeg frames which means they are can be over 100KB. Now all the major formats MP3, MP4, OGG, Opus and WMA should work

illuusio committed:
Tuning TryOpen to follow Mixxx coding guidelines

illuusio committed:
Trying to prevent crash to rule out 'imposible' situation that can't happen.

illuusio commented:
Ok last missing stuff is in. If there is no things that can be found this should be applied because it's uplifts FFmpeg to next level. After this is merge to master I'll port it to 1.12 so it'll also have good support for FFmpeg. You should try this with MP3/4, M4A, OGG, Opus and WMA (Pure AAC should work also).

daschuer commented on src/sources/soundsourceffmpeg.cpp:
```diff
@@ -415,28 +484,57 @@ bool SoundSourceFFmpeg::getBytesFromCache(char *buffer, SINT offset,
-                l_SObj = m_SCache[++ l_lPos];
-                continue;
+                if ((l_lPos + 1) < m_SCache.size())
+                {
```
Please move open braces to the end of line and use `} else {` below

daschuer commented:
Please try to break long lines before column 80, if possible.

illuusio committed:
Style branches at the end not the next line

illuusio committed:
Trying to keep 80 char maximum line length

illuusio commented:
Now they are fixed. There is still over 80 char lines but they are hard to separate so I'll keep them and if someone feels really bad I'll do something about them.

daschuer commented:
Is there something special for compiling with Ubuntu Trusty? 
Here is the tail of my console output 

```
src/encoder/encoderffmpegresample.cpp: In destructor ‘EncoderFfmpegResample::~EncoderFfmpegResample()’:
src/encoder/encoderffmpegresample.cpp:37:9: warning: ‘void audio_resample_close(ReSampleContext*)’ is deprecated (declared at /usr/include/libavcodec/avcodec.h:4126) [-Wdeprecated-declarations]
         audio_resample_close(m_pSwrCtx);
         ^
src/encoder/encoderffmpegresample.cpp:37:39: warning: ‘void audio_resample_close(ReSampleContext*)’ is deprecated (declared at /usr/include/libavcodec/avcodec.h:4126) [-Wdeprecated-declarations]
         audio_resample_close(m_pSwrCtx);
                                       ^
src/encoder/encoderffmpegresample.cpp: In member function ‘int EncoderFfmpegResample::open(AVSampleFormat, AVSampleFormat)’:
src/encoder/encoderffmpegresample.cpp:50:39: error: ‘AV_CH_LAYOUT_STEREO’ was not declared in this scope
         m_pCodecCtx->channel_layout = AV_CH_LAYOUT_STEREO;
                                       ^
src/encoder/encoderffmpegresample.cpp:52:39: error: ‘AV_CH_LAYOUT_MONO’ was not declared in this scope
         m_pCodecCtx->channel_layout = AV_CH_LAYOUT_MONO;
                                       ^
src/encoder/encoderffmpegresample.cpp:54:39: error: ‘AV_CH_LAYOUT_STEREO’ was not declared in this scope
         m_pCodecCtx->channel_layout = AV_CH_LAYOUT_STEREO;
                                       ^
src/encoder/encoderffmpegresample.cpp:68:44: error: ‘AV_CH_LAYOUT_STEREO’ was not declared in this scope
             m_pCodecCtx->channel_layout != AV_CH_LAYOUT_STEREO) && m_pSwrCtx == NULL) {
                                            ^
src/encoder/encoderffmpegresample.cpp:83:13: warning: ‘void audio_resample_close(ReSampleContext*)’ is deprecated (declared at /usr/include/libavcodec/avcodec.h:4126) [-Wdeprecated-declarations]
             audio_resample_close(m_pSwrCtx);
             ^
src/encoder/encoderffmpegresample.cpp:83:43: warning: ‘void audio_resample_close(ReSampleContext*)’ is deprecated (declared at /usr/include/libavcodec/avcodec.h:4126) [-Wdeprecated-declarations]
             audio_resample_close(m_pSwrCtx);
                                           ^
src/encoder/encoderffmpegresample.cpp:109:21: warning: ‘ReSampleContext* av_audio_resample_init(int, int, int, int, AVSampleFormat, AVSampleFormat, int, int, int, double)’ is deprecated (declared at /usr/include/libavcodec/avcodec.h:4109) [-Wdeprecated-declarations]
         m_pSwrCtx = av_audio_resample_init(m_pCodecCtx->channels,
                     ^
src/encoder/encoderffmpegresample.cpp:118:47: warning: ‘ReSampleContext* av_audio_resample_init(int, int, int, int, AVSampleFormat, AVSampleFormat, int, int, int, double)’ is deprecated (declared at /usr/include/libavcodec/avcodec.h:4109) [-Wdeprecated-declarations]
                                            0.8);
                                               ^
src/encoder/encoderffmpegresample.cpp: In member function ‘unsigned int EncoderFfmpegResample::reSample(AVFrame*, quint8**)’:
src/encoder/encoderffmpegresample.cpp:228:19: error: cannot convert ‘short int*’ to ‘quint8** {aka unsigned char**}’ in assignment
         outbuffer = (short *)malloc(l_iOutBytes * 2);
                   ^
src/encoder/encoderffmpegresample.cpp:257:18: warning: ‘int audio_resample(ReSampleContext*, short int*, short int*, int)’ is deprecated (declared at /usr/include/libavcodec/avcodec.h:4117) [-Wdeprecated-declarations]
         l_iLen = audio_resample(m_pSwrCtx,
                  ^
src/encoder/encoderffmpegresample.cpp:259:52: warning: ‘int audio_resample(ReSampleContext*, short int*, short int*, int)’ is deprecated (declared at /usr/include/libavcodec/avcodec.h:4117) [-Wdeprecated-declarations]
                                 inframe->nb_samples);
                                                    ^
src/encoder/encoderffmpegresample.cpp:190:16: warning: unused variable ‘l_lInReadBytes’ [-Wunused-variable]
         qint64 l_lInReadBytes = av_samples_get_buffer_size(NULL, m_pCodecCtx->channels,
                ^
scons: *** [lin64_build/encoder/encoderffmpegresample.o] Error 1
scons: building terminated because of errors.
```

illuusio commented:
Which version of avconv does it have?

daschuer commented:
Its is packed into libav-tools
6:9.18-0ubuntu0.14.04.1

illuusio commented:
Can't re-produce this.. what does '/usr/include/libavutil/channel_layout.h' contain?

daschuer commented:
After installing libavresample-dev it compiles without any issues. 
Is this an Ubuntu only issue? Can we add it to the build requirements?

daschuer commented:
I get a segfaults trying to play files
```
Debug [AnalyserQueue 1]: AnalyserWaveform::loadStored - Stored waveform loaded 
Debug [AnalyserQueue 1]: Beat calculation will not start 
Debug [AnalyserQueue 1]: Vampanalyser BlockSize:  32768 
Debug [AnalyserQueue 1]: Vampanalyser StepSize:  32768 
Debug [AnalyserQueue 1]: Key calculation started with plugin "qm-keydetector:2" 
Debug [AnalyserQueue 1]: EncoderFfmpegResample::reSample: nb_samples is zero 
Debug [AnalyserQueue 1]: SoundSourceFFmpeg::readFramesToCache: General error in audio decode: 0 
Debug [AnalyserQueue 1]: SoundSourceFFmpeg::readFramesToCache: Packet too big or File end 
Debug [AnalyserQueue 1]: SoundSourceFFmpeg::readFramesToCache: Packet too big or File end 
Fatal [AnalyserQueue 1]: ASSERT failure in QVector<T>::operator[]: "index out of range", file /usr/include/qt4/QtCore/qvector.h, line 359

Program received signal SIGABRT, Aborted.
[Switching to Thread 0x7fff83fff700 (LWP 18442)]
0x00007ffff0317cc9 in __GI_raise (sig=sig@entry=6)
    at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
56	../nptl/sysdeps/unix/sysv/linux/raise.c: Datei oder Verzeichnis nicht gefunden.
(gdb) bt
#0  0x00007ffff0317cc9 in __GI_raise (sig=sig@entry=6)
    at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007ffff031b0d8 in __GI_abort () at abort.c:89
#2  0x00007ffff53f5c92 in qt_message_output(QtMsgType, char const*) ()
   from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#3  0x00007ffff53f5ff9 in ?? () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#4  0x00007ffff53f6804 in qFatal(char const*, ...) ()
   from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#5  0x0000000000ad1473 in operator[] (i=-41, this=<optimized out>)
    at /usr/include/qt4/QtCore/qvector.h:359
#6  Mixxx::SoundSourceFFmpeg::getBytesFromCache (
    this=this@entry=0x7fff8800dc30, buffer=buffer@entry=0x3701b80, 
    offset=4672, size=size@entry=4096) at src/sources/soundsourceffmpeg.cpp:483
#7  0x0000000000ad19ed in Mixxx::SoundSourceFFmpeg::readSampleFrames (
    this=0x7fff8800dc30, numberOfFrames=4096, sampleBuffer=0x3701b80)
    at src/sources/soundsourceffmpeg.cpp:644
#8  0x0000000000acc260 in Mixxx::AudioSource::readSampleFramesStereo (
    this=0x7fff8800dc38, numberOfFrames=4096, sampleBuffer=0x3701b80, 
    sampleBufferSize=<optimized out>) at src/sources/audiosource.cpp:61
#9  0x00000000004c4ea8 in readSampleFramesStereo (numberOfFrames=4096, 
    pSampleBuffer=<optimized out>, this=0x7fff8800dc38)
    at src/sources/audiosource.h:206
#10 AnalyserQueue::doAnalysis (this=this@entry=0x3701910, tio=..., 
---Type <return> to continue, or q <return> to quit---
    pAudioSource=...) at src/analyserqueue.cpp:184
#11 0x00000000004caf8d in AnalyserQueue::run (this=0x3701910)
    at src/analyserqueue.cpp:334
#12 0x00007ffff540032f in ?? () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#13 0x00007ffff31e6182 in start_thread (arg=0x7fff83fff700)
    at pthread_create.c:312
#14 0x00007ffff03db47d in clone ()
    at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
(gdb) 
```

daschuer commented:
Here an other one: 

```
Debug [Main]: WSpinny::slotCoverFound WSpinny(0x5922980, name = "SpinnySingleton1")  "CoverInfo(NONE,GUESSED,,0,/home/daniel/workspace/t/libaudio-scan-perl-master/t/ogg/bug1155-2.ogg)" QSize(0, 0) 
Debug [AnalyserQueue 1]: ffmpeg: NEW FFMPEG API using libavresample 
Debug [AnalyserQueue 1]: Created sample rate converter for conversion of 44100 Hz format: fltp with: 2 (layout: 3 ) channels (BPS 4 ) 
Debug [AnalyserQueue 1]: To  44100  HZ format: flt with  2  (layout: 3 ) channels (BPS  4 ) 
Debug [AnalyserQueue 1]: SoundSourceFFmpeg::tryOpen: Samplerate:  44100 , Channels:  2 
 
Debug [AnalyserQueue 1]: AnalysisDAO fetched 2 analyses, 101488 bytes for track 4 in 4 ms 
Debug [AnalyserQueue 1]: Reading waveform from byte array: allSignalSize 8822 visualSampleRate 441 audioVisualRatio 100 
Debug [AnalyserQueue 1]: Reading waveform from byte array: allSignalSize 3842 visualSampleRate 192 audioVisualRatio 229.688 
Debug [AnalyserQueue 1]: AnalyserWaveform::loadStored - Stored waveform loaded 
Debug [AnalyserQueue 1]: Beat calculation will not start 
Debug [AnalyserQueue 1]: Vampanalyser BlockSize:  32768 
Debug [AnalyserQueue 1]: Vampanalyser StepSize:  32768 
Debug [AnalyserQueue 1]: Key calculation started with plugin "qm-keydetector:2" 
Debug [AnalyserQueue 1]: EncoderFfmpegResample::reSample: nb_samples is zero 
Debug [AnalyserQueue 1]: SoundSourceFFmpeg::readFramesToCache: General error in audio decode: 0 
Debug [CachingReaderWorker 1]: EncoderFfmpegResample::reSample: nb_samples is zero 
Debug [CachingReaderWorker 1]: SoundSourceFFmpeg::readFramesToCache: General error in audio decode: 0 
Debug [AnalyserQueue 1]: SoundSourceFFmpeg::readFramesToCache: Packet too big or File end 
Debug [AnalyserQueue 1]: SoundSourceFFmpeg::readFramesToCache: Packet too big or File end 
Debug [AnalyserQueue 1]: SoundSourceFFmpeg::readFramesToCache: Packet too big or File end 
Warning [AnalyserQueue 1]: DEBUG ASSERT: "framesRead <= framesToRead" in file src/analyserqueue.cpp, line 185
Warning [AnalyserQueue 1]: DEBUG ASSERT: "pAudioSource->isValidFrameIndex(frameIndex)" in file src/analyserqueue.cpp, line 187
Warning [AnalyserQueue 1]: DEBUG ASSERT: "pAudioSource->isValidFrameIndex(frameIndex)" in file src/analyserqueue.cpp, line 222
Debug [AnalyserQueue 1]: Key Detection complete 
Debug [AnalyserQueue 1]: Key Histogram 
Debug [AnalyserQueue 1]: ~SoundSourceFFmpeg(): Clear FFMPEG stuff 

Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fff87fff700 (LWP 18492)]
0x00007ffff10f24c3 in av_opt_next ()
   from /usr/lib/x86_64-linux-gnu/libavutil.so.52
(gdb) bt
#0  0x00007ffff10f24c3 in av_opt_next ()
   from /usr/lib/x86_64-linux-gnu/libavutil.so.52
#1  0x00007ffff10f2a4b in av_opt_free ()
   from /usr/lib/x86_64-linux-gnu/libavutil.so.52
#2  0x00007ffff13d63b2 in avformat_free_context ()
   from /usr/lib/x86_64-linux-gnu/libavformat.so.54
#3  0x0000000000acd355 in Mixxx::SoundSourceFFmpeg::close (
    this=this@entry=0x7fff8000d830) at src/sources/soundsourceffmpeg.cpp:164
#4  0x0000000000acd8b8 in ~SoundSourceFFmpeg (this=0x7fff8000d830, 
    __in_chrg=<optimized out>) at src/sources/soundsourceffmpeg.cpp:65
#5  Mixxx::SoundSourceFFmpeg::~SoundSourceFFmpeg (this=0x7fff8000d830, 
    __in_chrg=<optimized out>) at src/sources/soundsourceffmpeg.cpp:66
#6  0x00000000004cb97c in deref (value=0x7fff8000d830, d=0x7fff80005530)
    at /usr/include/qt4/QtCore/qsharedpointer_impl.h:342
#7  deref (this=0x7fff87ffedb8)
    at /usr/include/qt4/QtCore/qsharedpointer_impl.h:336
#8  ~ExternalRefCount (this=0x7fff87ffedb8, __in_chrg=<optimized out>)
    at /usr/include/qt4/QtCore/qsharedpointer_impl.h:401
#9  ~QSharedPointer (this=0x7fff87ffedb8, __in_chrg=<optimized out>)
    at /usr/include/qt4/QtCore/qsharedpointer_impl.h:466
#10 ~SoundSourceProxy (this=0x7fff87ffed90, __in_chrg=<optimized out>)
    at src/soundsourceproxy.h:11
#11 AnalyserQueue::run (this=0x37020d0) at src/analyserqueue.cpp:310
---Type <return> to continue, or q <return> to quit---
#12 0x00007ffff540032f in ?? () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#13 0x00007ffff31e6182 in start_thread (arg=0x7fff87fff700)
    at pthread_create.c:312
#14 0x00007ffff03db47d in clone ()
    at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
(gdb) 
```

illuusio commented:
Seems that libav have issues.. I'll fix those :dancer:

illuusio committed:
Fix compiling with libAV (Avconv) 0.8.x

illuusio committed:
Fix compiling on libav (Avconv) 9.18 without avresample

daschuer commented:
I van confirm that it no compiles without avresample. 
The sgfaults persists.

illuusio committed:
According to documentation: 'avformat_close_input' will free everything and set FormatCtx to NULL. In Avconv it crashes everything and in FFmpeg it pass. Correct way of letting evertyhing cleanup is just use 'avformat_close_input' alone

illuusio committed:
Old resampling API can't handle conversion from FLTP to FLT or S16P to S16 so make sure that user understands the situation

illuusio committed:
Correct FFMPEG to FFmpeg

illuusio committed:
Correct LibAV to Avconv (Which is prefered)

illuusio committed:
Correct the wording of FFmpeg and Avconv descriptions and update version info

illuusio commented:
Ok fixed that Avconv thing. Also make impossible to build Mixxx without having livavresample or libswresample starting from Avconv version 9.

daschuer commented:
The av_opt_free () segfault is gone. 
The QVector<T>::operator[]: Assertion is still there.

illuusio commented:
How do you get it? What file type?

daschuer commented:
> Old resampling API can't handle conversion from FLTP to FLT or S16P to S16 so make sure that user understands the situation

I still do not understand the situation ;-) 

If Mixxx requires avresample, we break Ubuntu Precise support which will last until 2017. 
Which Audio files are planar? 
Will it be reasonable to add a scons switch to disable planar support and allow to compile without avresample? 
Are there other aspects?

daschuer commented:
How do you get it? What file type?

I cannot play any file format e.g mp3.

illuusio commented:
Ok let's explain this Planar again. Mixxx Ffmpeg/Avconv implementation still supports Avconv 0.8.9 / FFmpeg 0.11-1.1 which doesn't use Planar  system. All format are converted to Planar in FFmpeg/Avconv after 1.2/9 and this means that old (which is depricated on Avconv 9 or FFmpeg 1.2 and above) resample API is out of question and only way to get from planar to non-planar (A.K.A PCM style) samples without some own written conversion methods (which I'm not willing to write because 9.18 have avresample which supports this conversion between FLTP --> FLT).
I don't know why we want to disable Planar (and I don't know if it's even possible) from Ubuntu repo packages when there is libavresample available?

daschuer commented:
Ah thank you, I am starting to understand. 

Ubuntu Precise has libav 0.8.17
Ubuntu Trusty has libav 9.18

From the 9.18 release notes: 
> In the libavcodec library, one of the most notable changes is added support for
planar audio (i.e. not interleaved). Many decoders and encoders, that previously
did inefficient (de)interleaving internally, now only work with planar audio
formats. Libavresample can be used for optimized conversion between interleaved
and planar formats.

For me it looks like that libav 0.8.17 does not output any planar formats. So there might be no reason for 
avresample on Ubuntu Precise. 

Can't we make the avresample requirement conditional for libav > 0.8.17? 
I think the old api code is already there, right?

illuusio commented:
As I said you can't use OLD API because at least in Avconv doesn't have needed conversion from FLTP (planar float) to FTL (normal float).  If you have some other converted that supports that conversion it can be made optional but I don't see the need.

illuusio committed:
Fix crashing on Avconv 9 and above

daschuer commented:
My theory is that it is a planar to interleave conversion is not required in the old API, because no codec has a planar output. 

Your comment here is a hint in that direction:  
https://github.com/mixxxdj/mixxx/blob/3f854c1674084609497036ba92e1605a43fb492f/src/encoder/encoderffmpegresample.cpp#L66

However, there are the planar enums are already defined in Precise's samplefmt.h 
They where introduced here: 
https://github.com/libav/libav/commit/8889cc4f5b767b323901115a92318a024336e2a1

An other thought: 
Are we sure Avconv is compiled with SSE enabled? 
If not, we may consider to use Mixxx's sample utils, which are SSE optimized. 
https://github.com/mixxxdj/mixxx/blob/3f854c1674084609497036ba92e1605a43fb492f/src/sampleutil.cpp#L341

illuusio commented:
With old 0.8.x samples are non-planar a.k.a interleave and everything works smoothly with no libavresample. 
Only thing missing from those converters are from planar stereo 16 to planar stereo Float.

illuusio committed:
Change resampler to use Mixxx inner functions and not FFmpeg library to convert sample formats to correct

illuusio committed:
Remove need for libavresampler/libswresampler or ffmpeg old resampler so it can be removed from depencies

illuusio committed:
Remove from depency for libavresample or libswresample (no need for FFmpeg old resampler) to favor Mixxx inner functions

illuusio committed:
Notify user that FFmpeg/Avconv has been enabled

illuusio committed:
Fix crashing in LibAV 9 and above

illuusio commented:
I should be so lucky lucky :8ball:. Now remove depencies for libswresample, libavresample and ffmpeg own resampler functions to favor Mixxx. There is no mono to stereo (is it needed to turn mono to stereo?) and no up/downsampling of samples.. hoping someone else is doing that job. If this is not what you wanted then I just give up and use this as my own and we can close this Pull Request

daschuer commented:
Cool Thank you. 

Since the new reSampleMixxx code assumes stereo, you should add a 
m_pCodecCtx->channels > 1 check. 

On the other hand, the mono path Is quite easy since there is no difference between planar and interleaved. 

Up/downsampling of sample rates should not be required, since the Mixxx engine will do the job. 
Just check, if the correct sample rate is reported.

daschuer commented on src/encoder/encoderffmpegresample.cpp:
```diff
@@ -17,262 +17,108 @@
+        return l_lOutReadBytes;
+
+    } else {
+        qDebug() << "Unknow sample format:" << av_get_sample_fmt_name(m_pCodecCtx->sample_fmt) << av_get_sample_fmt_name(m_pOutSampleFmt);
```
"Unsupported sample format:"

illuusio committed:
Case is more readable than if-else-if system for conversion

illuusio commented:
Even though last commit is little bit hustle (because I suck in GIT) this is it. Mono/Stereo plays on Avconv and FFmpeg. If there is not any other problems this should be merged that I can move on to next task. Encoders are not working but I understood they are not in use currently and they should be fixed in separate PR

illuusio committed:
Now Mono files also works

daschuer commented:
This does not merge cleanly with master. (only two tiny conflicts)

daschuer commented:
I have debug asserts with faulty ogg files. 

I think it was this https://github.com/jrmuizel/AudioScan-opus/blob/master/t/ogg/bug1155-2.ogg
(in this repro are some other interesting test candidates)

This is the relevant part of the log: 

```
Debug [AnalyserQueue 1]: New SoundSourceFFmpeg : "/home/daniel/workspace/t/libaudio-scan-perl-master/t/ogg/bug1155-2.ogg" 
Debug [AnalyserQueue 1]: EncoderFfmpegResample::openMixxx: open MIXXX FFmpeg Resampler version 
Debug [AnalyserQueue 1]: Created sample rate converter for conversion of 44100 Hz format: fltp with: 2 (layout: 3 ) channels (BPS 4 ) 
Debug [AnalyserQueue 1]: To  44100  HZ format: flt with  2  (layout: 3 ) channels (BPS  4 ) 
Debug [AnalyserQueue 1]: SoundSourceFFmpeg::tryOpen: Samplerate:  44100 , Channels:  2 
 
Debug [AnalyserQueue 1]: AnalyserWaveform::loadStored - Stored waveform loaded 
Debug [AnalyserQueue 1]: Beat calculation will not start 
Debug [AnalyserQueue 1]: Vampanalyser BlockSize:  32768 
Debug [AnalyserQueue 1]: Vampanalyser StepSize:  32768 
Debug [AnalyserQueue 1]: Key calculation started with plugin "qm-keydetector:2" 
Debug [AnalyserQueue 1]: EncoderFfmpegResample::reSample: nb_samples is zero 
Debug [AnalyserQueue 1]: SoundSourceFFmpeg::readFramesToCache: General error in audio decode: 0 
Debug [AnalyserQueue 1]: SoundSourceFFmpeg::readFramesToCache: Packet too big or File end 
Debug [AnalyserQueue 1]: SoundSourceFFmpeg::readFramesToCache: Packet too big or File end 
Debug [AnalyserQueue 1]: SoundSourceFFmpeg::readFramesToCache: Packet too big or File end 
Warning [AnalyserQueue 1]: DEBUG ASSERT: "framesRead <= framesToRead" in file src/analyserqueue.cpp, line 185
Warning [AnalyserQueue 1]: DEBUG ASSERT: "pAudioSource->isValidFrameIndex(frameIndex)" in file src/analyserqueue.cpp, line 187
Warning [AnalyserQueue 1]: DEBUG ASSERT: "pAudioSource->isValidFrameIndex(frameIndex)" in file src/analyserqueue.cpp, line 222
Debug [AnalyserQueue 1]: Key Detection complete 
Debug [AnalyserQueue 1]: Key Histogram 
Debug [AnalyserQueue 1]: ~SoundSourceFFmpeg(): Clear FFMPEG stuff 
Debug [AnalyserQueue 1]: ~SoundSourceFFmpeg(): Delete FFMPEG Resampler 
```

daschuer commented:
This file https://github.com/jrmuizel/AudioScan-opus/blob/master/t/flac/id3tagged.flac
makes Mixxx crash: 

```
Debug [CachingReaderWorker 2]: New SoundSourceFFmpeg : "/home/daniel/workspace/t/libaudio-scan-perl-master/t/flac/id3tagged.flac" 
Debug [CachingReaderWorker 2]: EncoderFfmpegResample::openMixxx: open MIXXX FFmpeg Resampler version 
Debug [CachingReaderWorker 2]: Created sample rate converter for conversion of 44100 Hz format: s16 with: 2 (layout: 3 ) channels (BPS 2 ) 
Debug [CachingReaderWorker 2]: To  44100  HZ format: flt with  2  (layout: 3 ) channels (BPS  4 ) 
Debug [CachingReaderWorker 2]: SoundSourceFFmpeg::tryOpen: Samplerate:  44100 , Channels:  2 
 
Debug [AnalyserQueue 1]: Vampanalyser BlockSize:  1024 
Debug [AnalyserQueue 1]: Vampanalyser StepSize:  512 
Debug [AnalyserQueue 1]: Beat calculation started with plugin "qm-tempotracker:0" 
Debug [AnalyserQueue 1]: Vampanalyser BlockSize:  32768 
Debug [AnalyserQueue 1]: Vampanalyser StepSize:  32768 
Debug [AnalyserQueue 1]: Key calculation started with plugin "qm-keydetector:2" 
Debug [AnalyserQueue 1]: SoundSourceFFmpeg::readFramesToCache: Packet too big or File end 
Fatal [AnalyserQueue 1]: ASSERT: "!isEmpty()" in file /usr/include/qt4/QtCore/qvector.h, line 260

Program received signal SIGABRT, Aborted.
[Switching to Thread 0x7fff83fff700 (LWP 24168)]
0x00007ffff0533cc9 in __GI_raise (sig=sig@entry=6)
    at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
56	../nptl/sysdeps/unix/sysv/linux/raise.c: Datei oder Verzeichnis nicht gefunden.
(gdb) bt
#0  0x00007ffff0533cc9 in __GI_raise (sig=sig@entry=6)
    at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007ffff05370d8 in __GI_abort () at abort.c:89
#2  0x00007ffff53f5c92 in qt_message_output(QtMsgType, char const*) ()
   from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#3  0x00007ffff53f5ff9 in ?? () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#4  0x00007ffff53f6804 in qFatal(char const*, ...) ()
   from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#5  0x0000000000ad28bc in first (this=0x7fff8c00de28)
    at /usr/include/qt4/QtCore/qvector.h:260
#6  Mixxx::SoundSourceFFmpeg::readFramesToCache (
    this=this@entry=0x7fff8c00dd90, count=count@entry=100, 
    offset=offset@entry=0) at src/sources/soundsourceffmpeg.cpp:398
#7  0x0000000000ad5fb8 in Mixxx::SoundSourceFFmpeg::seekSampleFrame (
    this=0x7fff8c00dd90, frameIndex=0) at src/sources/soundsourceffmpeg.cpp:631
#8  0x0000000000ad41f0 in Mixxx::SoundSourceFFmpeg::readSampleFrames (
    this=0x7fff8c00dd90, numberOfFrames=4096, sampleBuffer=0x370b9c0)
    at src/sources/soundsourceffmpeg.cpp:647
#9  0x0000000000ad0b70 in Mixxx::AudioSource::readSampleFramesStereo (
    this=0x7fff8c00dd98, numberOfFrames=4096, sampleBuffer=0x370b9c0, 
    sampleBufferSize=<optimized out>) at src/sources/audiosource.cpp:61
#10 0x00000000004c89d8 in readSampleFramesStereo (numberOfFrames=4096, 
    pSampleBuffer=<optimized out>, this=0x7fff8c00dd98)
---Type <return> to continue, or q <return> to quit---
    at src/sources/audiosource.h:206
#11 AnalyserQueue::doAnalysis (this=this@entry=0x370b880, tio=..., 
    pAudioSource=...) at src/analyserqueue.cpp:184
#12 0x00000000004ceabd in AnalyserQueue::run (this=0x370b880)
    at src/analyserqueue.cpp:334
#13 0x00007ffff540032f in ?? () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#14 0x00007ffff31e6182 in start_thread (arg=0x7fff83fff700)
    at pthread_create.c:312
#15 0x00007ffff05f747d in clone ()
    at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

daschuer commented:
An other one: 

```
Debug [CachingReaderWorker 2]: New SoundSourceFFmpeg : "/home/daniel/workspace/t/libaudio-scan-perl-master/t/flac/appId.flac" 
Debug [CachingReaderWorker 2]: EncoderFfmpegResample::openMixxx: open MIXXX FFmpeg Resampler version 
Debug [CachingReaderWorker 2]: Created sample rate converter for conversion of 44100 Hz format: s16 with: 2 (layout: 3 ) channels (BPS 2 ) 
Debug [CachingReaderWorker 2]: To  44100  HZ format: flt with  2  (layout: 3 ) channels (BPS  4 ) 
Debug [CachingReaderWorker 2]: SoundSourceFFmpeg::tryOpen: Samplerate:  44100 , Channels:  2 
 
Debug [CachingReaderWorker 2]: ~SoundSourceFFmpeg(): Clear FFMPEG stuff 
Debug [CachingReaderWorker 2]: ~SoundSourceFFmpeg(): Delete FFMPEG Resampler 
[flac @ 0x7fffa805b400] invalid sync code
[flac @ 0x7fffa805b400] invalid frame header
[flac @ 0x7fffa805b400] decode_frame() failed
Debug [CachingReaderWorker 2]: EOF! 
Fatal [CachingReaderWorker 2]: ASSERT: "!isEmpty()" in file /usr/include/qt4/QtCore/qvector.h, line 260

Program received signal SIGABRT, Aborted.
[Switching to Thread 0x7fffb5afc700 (LWP 24214)]
0x00007ffff0533cc9 in __GI_raise (sig=sig@entry=6)
    at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
56	../nptl/sysdeps/unix/sysv/linux/raise.c: Datei oder Verzeichnis nicht gefunden.
(gdb) bt
#0  0x00007ffff0533cc9 in __GI_raise (sig=sig@entry=6)
    at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007ffff05370d8 in __GI_abort () at abort.c:89
#2  0x00007ffff53f5c92 in qt_message_output(QtMsgType, char const*) ()
   from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#3  0x00007ffff53f5ff9 in ?? () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#4  0x00007ffff53f6804 in qFatal(char const*, ...) ()
   from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#5  0x0000000000ad28bc in first (this=0x7fffa80d2608)
    at /usr/include/qt4/QtCore/qvector.h:260
#6  Mixxx::SoundSourceFFmpeg::readFramesToCache (
    this=this@entry=0x7fffa80d2570, count=count@entry=100, 
    offset=offset@entry=0) at src/sources/soundsourceffmpeg.cpp:398
#7  0x0000000000ad5fb8 in Mixxx::SoundSourceFFmpeg::seekSampleFrame (
    this=0x7fffa80d2570, frameIndex=0) at src/sources/soundsourceffmpeg.cpp:631
#8  0x00000000004e35a9 in CachingReaderWorker::processChunkReadRequest (
    this=this@entry=0x21fde70, request=request@entry=0x7fffb5afbda0, 
    update=update@entry=0x7fffb5afbdc0) at src/cachingreaderworker.cpp:74
#9  0x00000000004e5872 in CachingReaderWorker::run (this=0x21fde70)
    at src/cachingreaderworker.cpp:140
#10 0x00007ffff540032f in ?? () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#11 0x00007ffff31e6182 in start_thread (arg=0x7fffb5afc700)
    at pthread_create.c:312
---Type <return> to continue, or q <return> to quit---
#12 0x00007ffff05f747d in clone ()
    at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```
```

daschuer commented:
There are some more crasher. Sorry for this hard test. I think all these crashes should not stop us from merging your work. It is still an optional feature and already an improvement. 
So if you running out of time working on this branch, we can postpone the work.

daschuer commented:
OK, it works good with normal files. Thank you for all the work. 
IMHO this can be merged after resolving the conflicts.

illuusio commented:
Nice crashers :D. I'll work them out like that merging issue.. Git makes me insane sometimes but I will manage.

illuusio committed:
Merge remote-tracking branch 'upstream/master' into ffmpeg-fixes

illuusio committed:
Now FFMPEG follows Mixxx code style and merges cleanly

illuusio committed:
Fixing playing BAD audio files:

    * https://github.com/jrmuizel/AudioScan-opus/blob/master/t/ogg/bug1155-2.ogg
    * https://github.com/jrmuizel/AudioScan-opus/blob/master/t/flac/id3tagged.flac

Doesn't crash anymore on empty cache. Probably needs more of those cache empty checks in future.

illuusio commented:
I that merge commit a problem? Should I revert once again and make it merge otherwise?

daschuer commented:
Sorry, I got an other one: 
 https://github.com/jrmuizel/AudioScan-opus/blob/master/t/flac/appId.flac

```
Debug [AnalyserQueue 1]: Keys version/sub-version unchanged since previous analysis. Not analyzing. 
Debug [AnalyserQueue 1]: Key calculation will not start. 
[flac @ 0x7fff885f37a0] invalid sync code
[flac @ 0x7fff885f37a0] invalid frame header
[flac @ 0x7fff885f37a0] decode_frame() failed
Debug [AnalyserQueue 1]: EOF! 
Debug [AnalyserQueue 1]: SoundSourceFFmpeg::readFramesToCache(): Can't read frames. Cache empty! 
Fatal [AnalyserQueue 1]: ASSERT failure in QVector<T>::operator[]: "index out of range", file /usr/include/qt4/QtCore/qvector.h, line 359

Program received signal SIGABRT, Aborted.
[Switching to Thread 0x7fff83fff700 (LWP 2695)]
0x00007ffff0533cc9 in __GI_raise (sig=sig@entry=6)
    at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
56	../nptl/sysdeps/unix/sysv/linux/raise.c: Datei oder Verzeichnis nicht gefunden.
(gdb) bt
#0  0x00007ffff0533cc9 in __GI_raise (sig=sig@entry=6)
    at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007ffff05370d8 in __GI_abort () at abort.c:89
#2  0x00007ffff53f5c92 in qt_message_output(QtMsgType, char const*) ()
   from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#3  0x00007ffff53f5ff9 in ?? () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#4  0x00007ffff53f6804 in qFatal(char const*, ...) ()
   from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#5  0x0000000000acfc2c in operator[] (i=0, this=0x7fff8803ac68)
    at /usr/include/qt4/QtCore/qvector.h:359
#6  Mixxx::SoundSourceFFmpeg::getBytesFromCache (
    this=this@entry=0x7fff8803abd0, buffer=buffer@entry=0x370cc50, offset=0, 
    size=size@entry=4096) at src/sources/soundsourceffmpeg.cpp:457
#7  0x0000000000ad05cd in Mixxx::SoundSourceFFmpeg::readSampleFrames (
    this=0x7fff8803abd0, numberOfFrames=4096, sampleBuffer=0x370cc50)
    at src/sources/soundsourceffmpeg.cpp:656
#8  0x0000000000accd00 in Mixxx::AudioSource::readSampleFramesStereo (
    this=0x7fff8803abd8, numberOfFrames=4096, sampleBuffer=0x370cc50, 
    sampleBufferSize=<optimized out>) at src/sources/audiosource.cpp:61
#9  0x00000000004c4b68 in readSampleFramesStereo (numberOfFrames=4096, 
    pSampleBuffer=<optimized out>, this=0x7fff8803abd8)
    at src/sources/audiosource.h:206
#10 AnalyserQueue::doAnalysis (this=this@entry=0x370cb10, tio=..., 
---Type <return> to continue, or q <return> to quit---
    pAudioSource=...) at src/analyserqueue.cpp:184
#11 0x00000000004cac4d in AnalyserQueue::run (this=0x370cb10)
    at src/analyserqueue.cpp:334
#12 0x00007ffff540032f in ?? () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#13 0x00007ffff31e6182 in start_thread (arg=0x7fff83fff700)
    at pthread_create.c:312
#14 0x00007ffff05f747d in clone ()
    at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

illuusio committed:
Fixing more bad audio file crashes:
   * https://github.com/jrmuizel/AudioScan-opus/blob/master/t/flac/appId.flac

Some stuff that should never happen but make program exit

illuusio commented:
Ok this should fix that.. Where these files comes from.. these are very buggy ones..

daschuer commented:
They are the test file from Audio::Scan found here: 
https://github.com/jrmuizel/AudioScan-opus

daschuer commented:
LGTM. Thank you very much for all the work!
